### PR TITLE
Execution Engine jobs are not invalidated automatically (#1268)

### PR DESCRIPTION
**Pull request by YaroslavTir**
Fixes https://github.com/OHDSI/WebAPI/issues/1268
- add scheduler class ExecutionEngineAnalysisStatusInvalidationScheduler, that is active only in case `execution.invalidation.maxage` and `execution.invalidation.period` are set
- small improvements

### CODE REVIEW

**Comment by YaroslavTir** [comment]
> When you click on FAILED status of invalidated PLP execution, popup doesn't arise. Should be
> image

indeed, Only status was invalidated, but not a job. I fixed it with the last commit.

**Comment by pavgra** [comment]
1. First, scheduler calls `invalidateExecutions -> jobInvalidator.invalidateJobExecutionById -> jobRepository.update`
2. `exit_code = Failed`  together with `exit_message = Invalidated by system` are properly persisted in database
3. Breakpoint in `org.springframework.batch.core.job.AbstractJob` line 341 (`finally` section of job execution) show exitStatus with empty message

File: src/main/java/org/ohdsi/webapi/JobInvalidator.java [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
 import org.springframework.stereotype.Component;
 import org.springframework.transaction.support.TransactionTemplate;
 
-import javax.annotation.PostConstruct;
-import java.util.Calendar;
-
 @Component
 public class JobInvalidator {
 
+    public static final String INVALIDATED_BY_SYSTEM_EXIT_CODE = "Invalidated by system";
```
Comment (chrisknoll) [inline src/main/java/org/ohdsi/webapi/JobInvalidator.java]: Is this a 'code' or a message?

Comment (pavgra) [reply src/main/java/org/ohdsi/webapi/JobInvalidator.java]: Message. So I agree, would be good to rename the variable

File: pom.xml [hunk 2]
```diff
@@ -143,6 +143,7 @@
     <arachne.version>1.15.0-SNAPSHOT</arachne.version>
     <jersey-media-multipart.version>2.25.1</jersey-media-multipart.version>
     <execution.invalidation.period>600000</execution.invalidation.period>
+    <execution.invalidation.maxage>4</execution.invalidation.maxage>
```
Comment (wivern) [inline pom.xml]: Could we provide some note about which units should be used for maxage?

Comment (wivern) [reply pom.xml]: As to me the 4 hours is to low to set as default, especially for estimation analysis. Could we increase it to 12 at least?

File: src/main/java/org/ohdsi/webapi/executionengine/service/ScriptExecutionServiceImpl.java [hunk 3]
```diff
@@ -81,7 +79,15 @@
...
 
-    private List<ExecutionEngineAnalysisStatus.Status> FINAL_STATUES = ImmutableList.of(ExecutionEngineAnalysisStatus.Status.COMPLETED, ExecutionEngineAnalysisStatus.Status.COMPLETED);
+    private static List<ExecutionEngineAnalysisStatus.Status> INVALIDATE_STATUSES = ImmutableList.of(
+            ExecutionEngineAnalysisStatus.Status.RUNNING,
+            ExecutionEngineAnalysisStatus.Status.STARTED,
+            ExecutionEngineAnalysisStatus.Status.PENDING
+    );
+
+    private static List<ExecutionEngineAnalysisStatus.Status> FINAL_STATUES = ImmutableList.of(
+            ExecutionEngineAnalysisStatus.Status.COMPLETED,
```
Comment (wivern) [inline src/main/java/org/ohdsi/webapi/executionengine/service/ScriptExecutionServiceImpl.java]: I guess it should be FAILED here

File: src/main/java/org/ohdsi/webapi/JobInvalidator.java [hunk 4]
```diff
@@ -29,15 +32,23 @@ public JobInvalidator(JobExplorer explorer, JobRepository repository, Transactio
...
             return null;
         });
     }
+
+    public void invalidateJobExecutionById(ExecutionEngineAnalysisStatus executionEngineAnalysisStatus) {
+        JobExecution job = jobExplorer.getJobExecution(executionEngineAnalysisStatus.getExecutionEngineGeneration().getId());
+        invalidationJobExecution(job);
+    }
+
+    private void invalidationJobExecution(JobExecution job) {
```
Comment (wivern) [inline src/main/java/org/ohdsi/webapi/JobInvalidator.java]: This is similar to the `CcServiceImpl.invalidateGenerations`, could we provide some code reuse?
E.g. via `AbstractDaoService`

File: pom.xml (hunk 2 above)
Comment (YaroslavTir) [reply pom.xml]: Good point, this value used to be 4 hours, so I just return it.

File: src/main/java/org/ohdsi/webapi/JobInvalidator.java (hunk 4 above)
Comment (YaroslavTir) [reply src/main/java/org/ohdsi/webapi/JobInvalidator.java]: good catch.

**Pull request merged**